File: 11/list.11.1-2.swift

import Foundation

// 社内の顧客管理システムとする

extension DateFormatter {
    convenience init(dateFormat: String) {
        self.init()
        self.dateFormat = dateFormat
    }
}

extension String {
    var date: Date {
        DateFormatter(dateFormat: "yyyy-MM-dd").date(from: self) ?? Date()
    }
}

extension Date {
    func check(byAdding component: Calendar.Component, value: Int) -> Bool {
        self < Calendar.current.date(byAdding: component, value: value, to: Date())!
    }
}

// 顧客
struct User {
    let id: Int
    let name: String
    var deletedAt: Date?
}

// 社員
struct Employee {
    let id: Int
    let name: String
    var createdAt: Date
    var deletedAt: Date?

    var intern: Bool
    
    // インターンが間違って優良ユーザを削除してしまった！
    // 顧客を削除できる社員を限定しよう！
    // 仕様変更に合わせてコメントを更新

    /*
     [ここのコメントが話の中心]
     インターン以外の社員が顧客を削除する
     */
    func delete(user: User) {
        guard !intern else {
          print("\(name)はユーザを削除する権限がありません")
          return
        }

        var deletedUser = user
        deletedUser.deletedAt = Date()
        // apiを叩いて登録
        print("\(name)はユーザを削除しました")
    }
}

// 本当はDBからもらうデータ
let blackUser = User(id: 1, name: "他からクレームがきまくったユーザ")

let employee1 = Employee(id: 1, name: "あなた", createdAt: "2000-10-01".date, intern: false)
let employee2 = Employee(id: 2, name: "インターン", createdAt: "2010-10-01".date, intern: true)

employee1.delete(user: blackUser)
employee2.delete(user: blackUser)

File: 11/list.11.1-2.dart
import 'package:intl/intl.dart';

// 社内の顧客管理システムとする

extension StringExtension on String {
  DateTime get date {
    return DateFormat('yyyy-MM-dd').parse(this);
  }
}

extension DateTimeExtension on DateTime {
  bool check({required DateTimeComponent component, required int value}) {
    return this
        .isBefore(DateTime.now().add(_durationFromComponent(component, value)));
  }
}

// DartにはCalendar.Componentがないため、独自の列挙型を作成
enum DateTimeComponent { day, month, year }

Duration _durationFromComponent(DateTimeComponent component, int value) {
  switch (component) {
    case DateTimeComponent.day:
      return Duration(days: value);
    case DateTimeComponent.month:
      return Duration(days: value * 30); // 簡易計算
    case DateTimeComponent.year:
      return Duration(days: value * 365); // 簡易計算
  }
}

// 顧客
class User {
  final int id;
  final String name;
  DateTime? deletedAt;

  User({required this.id, required this.name, this.deletedAt});
}

// 社員
class Employee {
  final int id;
  final String name;
  final DateTime createdAt;
  DateTime? deletedAt;

  final bool intern;

  Employee(
      {required this.id,
      required this.name,
      required this.createdAt,
      required this.intern,
      this.deletedAt});

  // インターンが間違って優良ユーザを削除してしまった！
  // 顧客を削除できる社員を限定しよう！
  // 仕様変更に合わせてコメントを更新

  /*
   [ここのコメントが話の中心]
   社員が顧客を削除する
   */
  void delete(User user) {
    if (intern) {
      print("$nameはユーザを削除する権限がありません");
      return;
    }
    user.deletedAt = DateTime.now();
    // apiを叩いて登録
    print("$nameはユーザを削除しました");
  }
}

void main() {
  // 本当はDBからもらうデータ
  var blackUser = User(id: 1, name: "他からクレームがきまくったユーザ");
  var employee1 =
      Employee(id: 1, name: "あなた", createdAt: "2000-10-01".date, intern: false);
  var employee2 = Employee(
      id: 2, name: "インターン", createdAt: "2010-10-01".date, intern: true);

  employee1.delete(blackUser);
  employee2.delete(blackUser);
}
